ravenml/data/write_dataset.py

```diff
@@ -0,0 +1,363 @@
+from ravenml.data.helpers import default_filter, copy_data_locally, split_data, read_json_metadata
+
+class DecoratorSuperClass:
+    """Superclass for DatasetWriter. Allows for decorators
+        from one class to be inherited to all subclasses. Subclasses can 
+        overload their decorators if desired.
+
+        Any decorator that is to be used on subclasses of this class must
+        set the attribute 'inherit_decorator' to the decorator function itself
+        in the inner function that it is passing. 
+
+        This class supports decorators with parameters, but requires that they
+        be set as attributes of the passed function as 'args' and 'kwargs'.
+    """
+    def __init_subclass__(cls):
+        decorator_registry = getattr(cls, "_decorator_registry", {}).copy()
+        cls._decorator_registry = decorator_registry
+        # annotate newly decorated methods in the current subclass:
+        for name, obj in cls.__dict__.items():
+            if getattr(obj, "inherit_decorator", False) and not name in decorator_registry:
+                decorator_registry[name] = (obj.inherit_decorator, getattr(obj, "args", False), getattr(obj, "kwargs", False))
+        # decorate methods annotated in the registry
+        # decorator[0] = decorator function
+        # decorator[1] = decorator args
+        # decorator[2] = decorator kwargs
+        for name, decorator in decorator_registry.items():
+            if name in cls.__dict__ and getattr(getattr(cls,name), "inherit_decorator", None) != decorator[0]:
+                if decorator[1] or decorator[2]:
+                    if decorator[2]:
+                        setattr(cls, name, decorator[0](decorator[1], decorator[2])(cls.__dict__[name]))
+                    else:
+                        setattr(cls, name, decorator[0](decorator[1])(cls.__dict__[name]))
+                elif not decorator[1]:
+                    setattr(cls, name, decorator[0](decorator[2])(cls.__dict__[name]))
+                else:
+                    setattr(cls, name, decorator[0](cls.__dict__[name]))
+
+class DatasetWriter(DecoratorSuperClass):
+    """Interface for creating datasets, methods are in order of what is expected to be 
+        called by the plugins
+
+    Methods:
+        __init__ (CreateInput): takes in CreateInput object and initializes variables
+            to be used by the rest of the methods and the plugins
+        load_image_ids (): gets all image_ids/tags from the supplied imagesets based on
+            finding metadata files
+        interactive_filter (): takes the current image_ids and allows the user to create
+            sets through interactive filtering using the image_ids
+        load_data (): copies all related files to the image_ids into a temp folder
+        construct_all (): plugin specific method to generate objects which will be used
+            in writing the dataset
+        write_dataset (obj_list): main driver for writing the dataset locally
```

Comment:
Why do all the other methods rely on setting instance variables, but then `write_dataset` requires an argument? I figure may as well standardize it so that they all use the same paradigm.